Implement people search by graduation year

antoniavonto opened the pull request:
As seen below, we have a visual representation of being able to search alumni based on the graduation year. For this to work, we also made changes to the backend API. We have been able to implement search based on graduation year but haven't been able to implement searching based on a range of years (potentially a slider). See: #1287 

![image](https://github.com/gordon-cs/gordon-360-ui/assets/115328139/f9b17952-51bb-4f67-8b9c-85374ccc8fb4)

antoniavonto commented:
See: https://github.com/gordon-cs/gordon-360-api/pull/913

bennettforkner commented:
Why is this a separate search field from "Class Year"? Are they not the same thing?

I imagine we should be defaulting to preferred no matter what anyway.

bennettforkner commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -80,13 +81,14 @@ const searchPageTitle = (
 const defaultSearchParams: PeopleSearchQuery = {
   includeStudent: true,
   includeFacStaff: true,
-  includeAlumni: false,
+  includeAlumni: true,
```
I think this should stay as false. I get that it's confusing that you need to check this and also enter a year, but we want to show only current community members unless the user requests alumni.

bennettforkner commented on .env.development:
```diff
@@ -7,4 +7,4 @@
 
 # @LOCALHOST
-# VITE_API_URL=http://localhost:51626/
+# VITE_API_URL=http://localhost:51631/
```
Please revert this change before merge.

bennettforkner reviewed (commented):
> I have changed the naming before seeing your previous message about the name "Academic year". I changed from "class year" to "class standing" and from "preferred graduation year" to "graduation year." What would be the most clarifying?

I like "class standing" better. Cool!

I still think it should all be within the Student/Alumni section though.

antoniavonto commented:
We added the switch component to toggle between the two different search options based on graduation year or a range of grad years. We also made changes to the API to make the range of graduation years to work. There are still some things that we would like to improve for this project before finalizing it.  
![image](https://github.com/gordon-cs/gordon-360-ui/assets/115328139/198e690f-40a6-4fb3-9a73-735997fced33)
![image](https://github.com/gordon-cs/gordon-360-ui/assets/115328139/ecee8645-b678-4c92-8803-5d96504b28ca)

bennettforkner commented:
> We added the switch component to toggle between the two different search options based on graduation year or a range of grad years. We also made changes to the API to make the range of graduation years to work. There are still some things that we would like to improve for this project before finalizing it.

Since this is already visually a switch, I would just call it "Search by Year Range"

bennettforkner commented:
Also can you please address my suggestion that this all go under a single Student/Alumni section? `Student/Alumni` is one section now. You don't see that as the header because you are of person type "student":
![image](https://github.com/gordon-cs/gordon-360-ui/assets/43764277/c19de1b1-4921-461e-aa27-9f1acf2654c2)

antoniavonto commented:
We have updated/changed all necessary suggestions we received. We fixed the bug so that someone would not be able to search for the class standing and graduation year parameters at the same time. The images below are the updated visual changes. 
![image](https://github.com/gordon-cs/gordon-360-ui/assets/115328139/f1b3b1c5-e726-4c71-b364-f70825a92ceb)
![image](https://github.com/gordon-cs/gordon-360-ui/assets/115328139/0169cb46-a79b-41c3-9b72-5fbde3078e94)
for mobile:
![image](https://github.com/gordon-cs/gordon-360-ui/assets/115328139/5d7a58bd-21e0-4874-a5ef-b0cc348b7b99)

amos-cha commented:
@antoniavonto your merge with develop created duplicate services. Fix this before we can proceed with reviews!

EjPlatzer commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -139,6 +150,13 @@ const SearchFieldList = ({ onSearch }: Props) => {
   const [departments, setDepartments] = useState<string[]>([]);
   const [buildings, setBuildings] = useState<string[]>([]);
   const [halls, setHalls] = useState<string[]>([]);
+  const [userProvidedYear, setUserProvidedYear] = useState(new Date().getFullYear());
```
This doesn't need to be a state variable, since it's never updated. You can make it a constant outside the component.

EjPlatzer replied:
Also, it's probably clearer to name it `currentYear`, since this variable is not actually user provided.

antoniavonto replied:
Yes that is true, thank you! Just changed that

EjPlatzer commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -139,6 +150,13 @@ const SearchFieldList = ({ onSearch }: Props) => {
   const [buildings, setBuildings] = useState<string[]>([]);
   const [halls, setHalls] = useState<string[]>([]);
+  const [userProvidedYear, setUserProvidedYear] = useState(new Date().getFullYear());
+  const [graduationYearRange, setGraduationYearRange] = useState<number[]>([
+    1889,
+    userProvidedYear,
+  ]);
```
I think that this doesn't need to be it's own state value, since we're already storing the initial year and final year in the search params. You can just provide defaults when the search param values are empty strings.

antoniavonto replied:
I could, instead of unsing the set function to change graduationYearRange to just separate straight the newValue, but then it would be hard to implement the visualization of the hands of the slider changing.
![image](https://github.com/gordon-cs/gordon-360-ui/assets/115328139/944b05f0-80e1-4adf-a116-cda2eb8c2d65)
![image](https://github.com/gordon-cs/gordon-360-ui/assets/115328139/b6ad7ebc-bb90-4a8e-b2ca-3615429be6dc)

russtuck replied:
@EjPlatzer are you satisfied with leaving this as it is?

EjPlatzer commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -421,6 +481,49 @@ const SearchFieldList = ({ onSearch }: Props) => {
+                      </Typography>
+                    </Grid>
+                  )}
+                  {switchYearRange == true ? (
+                    <FormControlLabel
+                      control={<Switch onChange={handleSwitchChange}></Switch>}
+                      label="Search by Year Range"
+                      labelPlacement="end"
+                    ></FormControlLabel>
+                  ) : (
+                    <FormControlLabel
+                      control={<Switch onChange={handleSwitchChange}></Switch>}
+                      label="Search by Graduation Year"
+                      labelPlacement="end"
+                    ></FormControlLabel>
+                  )}
```
We can condense this conditional case - since the FormControlLabel is always rendered and it's content is almost exactly the same, except the label:
```suggestion
                 <FormControlLabel
                      control={<Switch onChange={handleSwitchChange} />}
                      label={switchYearRange ? "Search by Year Range" : "Search by Graduation Year"}
                      labelPlacement="end"
                   />
```

antoniavonto replied:
That is true, it would make it more concise. Just changed it, thank you!

amos-cha commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -85,12 +92,16 @@ const defaultSearchParams: PeopleSearchQuery = {
   country: '',
   department: '',
   building: '',
+  //graduation_year_range: '',
```
```suggestion
  //graduation_year_range: '',
```
Is this meant to be commented out?

antoniavonto replied:
I forgot to remove this comment. Thanks for pointing it out!

amos-cha commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -139,6 +150,13 @@ const SearchFieldList = ({ onSearch }: Props) => {
   const [buildings, setBuildings] = useState<string[]>([]);
   const [halls, setHalls] = useState<string[]>([]);
+  const [userProvidedYear, setUserProvidedYear] = useState(new Date().getFullYear());
+  const [graduationYearRange, setGraduationYearRange] = useState<number[]>([
+    1889,
+    userProvidedYear,
+  ]);
```
```suggestion
  const [graduationYearRange, setGraduationYearRange] = useState<number[]>([
    1889,
    userProvidedYear,
  ]);
```

why 1889?

antoniavonto replied:
was the year the school was founded

EjPlatzer commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -421,6 +481,49 @@ const SearchFieldList = ({ onSearch }: Props) => {
+                    />
+                  ) : (
+                    <Grid item width={225}>
+                      <Slider
```
I like the idea of a slider here, but I'm concerned it may be the wrong choice in terms of UX. Since the range is several hundred years, it takes a lot of effort to get the value exactly what you want it to be. Consider someone on a touch screen who has slightly shaky hands - they would probably have great difficult setting this to be exactly 1970-1974, for example. I think in the interest of UX, we might need to change this to two separate year inputs.

antoniavonto replied:
I agree with you the slider is probably not the best option for this feature. We were thinking of some options and wanted your opinion. 

1. Keeping the slide, but also allowing some other way of changing the year range.
2. Making a select that allow you to choose two dates for the range
3. Making two text boxes for the initial and end year (Not my preferred option since people might type other things and it might not be intuitive for the user)

amos-cha replied:
I honestly don't know if there is a really good option. I think in terms of mobile usability I would argue that a location where we could select a year would be optimal. However, I don't know if the practicality of a year selector is much better (if any better) than the live updating of the slider.

amos-cha commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -245,19 +263,61 @@ const SearchFieldList = ({ onSearch }: Props) => {
+  };
+
+  const handleSwitchChange = () => {
+    if (switchYearRange === true) {
+      setSearchParams((sp) => ({
+        ...sp,
+        ...{ initial_year: '', final_year: '' },
+      }));
+    } else {
+      setSearchParams((sp) => ({
+        ...sp,
+        ...{ initial_year: '', final_year: '' },
+      }));
+    }
```
Is there a reason the if/else block has the exact same block of code?

antoniavonto replied:
No reason just mistake, just removed after testing it.

amos-cha commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -421,6 +481,49 @@ const SearchFieldList = ({ onSearch }: Props) => {
+                      name="graduation_year"
+                      value={searchParams.graduation_year}
+                      updateValue={handleUpdate}
+                      options={Array.from({ length: userProvidedYear - 1889 + 1 }, (_, i) => ({
```
again why 1889?

antoniavonto replied:
It is the year the school was founded

EjPlatzer reviewed (changes requested):
This is awesome! I have several notes on how the code can be cleaned up. The most notable change I'm recommending is that the `Slider` component is probably not ideal for this use case, although I like the idea.

antoniavonto commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -85,12 +92,16 @@ const defaultSearchParams: PeopleSearchQuery = {
   country: '',
   department: '',
   building: '',
+  //graduation_year_range: '',
```
I forgot to remove this comment, thanks for pointing it out!

antoniavonto commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -245,19 +263,61 @@ const SearchFieldList = ({ onSearch }: Props) => {
   }, [location.search, initialSearchParams]);
 
-  const handleUpdate = (event: ChangeEvent<HTMLInputElement>) =>
+  const handleUpdate = (event: ChangeEvent<HTMLInputElement>) => {
+    if (event.target.name === 'graduation_year') {
+      setSearchParams((sp) => ({
+        ...sp,
+        ...{ class_standing: '' },
+      }));
+    } else if (event.target.name === 'class_standing') {
+      setSearchParams((sp) => ({
+        ...sp,
+        ...{ initial_year: '', final_year: '' },
+        ...{ graduation_year: '' },
+      }));
+    }
+
     setSearchParams((sp) => ({
       ...sp,
       [event.target.name]:
         event.target.type === 'checkbox' ? event.target.checked : event.target.value,
     }));
+  };
```
Did not know about that, thank you!

antoniavonto commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -149,7 +167,7 @@ const SearchFieldList = ({ onSearch }: Props) => {
       includeStudent: isStudent || isFacStaff,
       // Only alumni search alumni by default
-      includeAlumni: isAlumni,
+      // includeAlumni: isAlumni,
```
I am not quite sure, but if not commented out the checkbox for including searching for alumni will be set checked as the default.

antoniavonto commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -139,6 +150,13 @@ const SearchFieldList = ({ onSearch }: Props) => {
+    userProvidedYear,
+  ]);
+  const [switchYearRange, setSwitchYearRange] = useState(true);
+  const [studentSearch, setStudentSearch] = useState(Boolean);
```
Not actually used studentSearch, just removed it

antoniavonto commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -245,19 +263,61 @@ const SearchFieldList = ({ onSearch }: Props) => {
   };
 
+  const handleSliderChange = (event: Event, newValue: number | number[]) => {
+    setGraduationYearRange(newValue as number[]);
```
When I try making it as:
![image](https://github.com/gordon-cs/gordon-360-ui/assets/115328139/1d9b55e6-e77c-4f7d-83f4-edc4dd23806d)
I get an error with the on change of the slider that says: Type 'number' is not assignable to type 'number[ ]'. This leads me to assume that newValue might not be always a number[ ], but I cannot see why.

antoniavonto replied:
I will not get too into this issue since we are thinking of moving away from the use of the slide

antoniavonto replied:
@EjPlatzer Since we created a new issue for finding a new solution in the place of the slider, is it ok to leave this like this?

amos-cha commented on .env.development:
```diff
@@ -8,3 +8,4 @@
 
 # @LOCALHOST
 # VITE_API_URL=http://localhost:51626/
+ 
```
```suggestion
```

amos-cha commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -149,7 +162,7 @@ const SearchFieldList = ({ onSearch }: Props) => {
       includeStudent: isStudent || isFacStaff,
       // Only alumni search alumni by default
-      includeAlumni: isAlumni,
+      // includeAlumni: isAlumni,
```
reminder to uncomment or remove

amos-cha commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -421,6 +474,41 @@ const SearchFieldList = ({ onSearch }: Props) => {
+                      name="graduation_year"
+                      value={searchParams.graduation_year}
+                      updateValue={handleUpdate}
+                      options={Array.from({ length: currentYear - 1889 + 1 }, (_, i) => ({
+                        value: (currentYear - i).toString(),
```
```suggestion
                      options={Array.from({ length: currentYear - 1889 + 1 }, (_, i) => ({
                        value: (currentYear - i).toString(),
```

I would suggest making a note somewhere that 1889 is the establish date of Gordon (thus earliest possible year)

amos-cha reviewed (dismissed):
Looks good, I think it's best to leave figuring out the UX as an issue. I don't think it is a necessarily blocking change.

russtuck commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -40,12 +41,18 @@ import {
+
+function valuetext(value: number) {
+  return '${value}';
+}
```
@antoniavonto this looks like it's still a to-do item

russtuck replied:
Nevermind, it looks done.  Sorry.

russtuck commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -245,19 +263,61 @@ const SearchFieldList = ({ onSearch }: Props) => {
 
+  const handleSliderChange = (event: Event, newValue: number | number[]) => {
+    setGraduationYearRange(newValue as number[]);
+    var values = graduationYearRange.toString().split(',');
```
@antoniavonto Have you addressed and tested the reliability concern?

russtuck commented:
It looks like this is close.  I commented on what appear to be the remaining issues.

antoniavonto commented:
@amos-cha

amos-cha commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -415,6 +465,41 @@ const SearchFieldList = ({ onSearch }: Props) => {
                     select
                     disabled={!searchParams.includeStudent}
                   />
+                  {switchYearRange == true ? (
```
```suggestion
                  {switchYearRange ? (
```

amos-cha commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -239,19 +249,59 @@ const SearchFieldList = ({ onSearch }: Props) => {
+  };
+
+  const handleSwitchChange = () => {
+    if (switchYearRange === true) {
```
```suggestion
    if (switchYearRange) {
```

amos-cha reviewed (changes requested):
Almost, then it looks good

amos-cha reviewed (approved):
👍🏼

antoniavonto commented:
@amos-cha

antoniavonto commented on src/views/PeopleSearch/components/SearchFieldList/index.tsx:
```diff
@@ -260,6 +260,7 @@ const SearchFieldList = ({ onSearch }: Props) => {
       setSearchParams((sp) => ({
         ...sp,
         class_standing: '',
+        minor: '',
```
Alumni's Minors are not saved, so the results would always be empty if graduation year was searched together with a minor

antoniavonto replied:
Alumni's Minors are not saved, so the results would always be empty if graduation year was searched together with a minor
